File: messagebus/src/main/java/com/yahoo/messagebus/MessageBus.java
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.messagebus;

import com.yahoo.concurrent.CopyOnWriteHashMap;
import com.yahoo.concurrent.SystemTimer;
import com.yahoo.log.LogLevel;
import com.yahoo.messagebus.metrics.MessageBusMetricSet;
import com.yahoo.messagebus.network.Network;
import com.yahoo.messagebus.network.NetworkOwner;
import com.yahoo.messagebus.routing.*;
import com.yahoo.text.Utf8Array;
import com.yahoo.text.Utf8String;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;
import java.util.logging.Logger;

/**
 * <p>A message bus contains the factory for creating sessions to send, receive
 * and forward messages.</p>
 *
 * <p>There are three types of sessions:</p>
 * <ul>
 *     <li>{@link SourceSession Source sessions} sends messages and receives replies</li>
 *     <li>{@link IntermediateSession Intermediate sessions} receives messages on
 *         their way to their final destination, and may decide to forward the messages or reply directly.
 *     <li>{@link DestinationSession Destination sessions} are the final recipient
 *         of messages, and are expected to reply to every one of them, but may not forward messages.
 * </ul>
 *
 * <p>A message bus is configured with a {@link Protocol protocol}. This table
 * enumerates the permissible routes from intermediates to destinations and the
 * messaging semantics of each hop.</p>
 *
 * The responsibilities of a message bus are:
 * <ul>
 *     <li>Assign a route to every send message from its routing table
 *     <li>Deliver every message it <i>accepts</i> to the next hop on its route
 *         <i>or</i> deliver a <i>failure reply</i>.
 *     <li>Deliver replies back to message sources through all the intermediate hops.
 * </ul>
 *
 * A runtime will typically
 * <ul>
 *     <li>Create a message bus implementation and set properties on this implementation once.
 *     <li>Create sessions using that message bus many places.</li>
 * </ul>
 *
 * @author bratseth
 * @author Simon Thoresen
 */
public class MessageBus implements ConfigHandler, NetworkOwner, MessageHandler, ReplyHandler {

    private static Logger log = Logger.getLogger(MessageBus.class.getName());
    private final AtomicBoolean destroyed = new AtomicBoolean(false);
    private final ProtocolRepository protocolRepository = new ProtocolRepository();
    private final AtomicReference<Map<String, RoutingTable>> tablesRef = new AtomicReference<Map<String, RoutingTable>>(null);
    private final CopyOnWriteHashMap<String, MessageHandler> sessions = new CopyOnWriteHashMap<String, MessageHandler>();
    private final Network net;
    private final Messenger msn;
    private final Resender resender;
    private int maxPendingCount = 0;
    private int maxPendingSize = 0;
    private int pendingCount = 0;
    private int pendingSize = 0;
    private final Thread careTaker = new Thread(this::sendBlockedMessages);
    private final ConcurrentHashMap<SendBlockedMessages, Long> blockedSenders = new ConcurrentHashMap<>();
    private MessageBusMetricSet metrics = new MessageBusMetricSet();

    public interface SendBlockedMessages {
        /**
         * Do what you want, but dont block.
         * You will be called regularly until you signal you are done
         * @return true unless you are done
         */
        boolean trySend();
    }

    public void register(SendBlockedMessages sender) {
        blockedSenders.put(sender, SystemTimer.INSTANCE.milliTime());
    }

    private void sendBlockedMessages() {
        while (! destroyed.get()) {
            for (SendBlockedMessages sender : blockedSenders.keySet()) {
                if (!sender.trySend()) {
                    blockedSenders.remove(sender);
                }
            }
            try {

                Thread.sleep(10);
            } catch (InterruptedException e) {
                return;
            }
        }
    }

    /**
     * <p>Convenience constructor that proxies {@link #MessageBus(Network,
     * MessageBusParams)} by adding the given protocols to a default {@link
     * MessageBusParams} object.</p>
     *
     * @param net       The network to associate with.
     * @param protocols An array of protocols to register.
     */
    public MessageBus(Network net, List<Protocol> protocols) {
        this(net, new MessageBusParams().addProtocols(protocols));
    }

    /**
     * <p>Constructs an instance of message bus. This requires a network object
     * that it will associate with. This assignment may not change during the
     * lifetime of this message bus.</p>
     *
     * @param net    The network to associate with.
     * @param params The parameters that controls this bus.
     */
    public MessageBus(Network net, MessageBusParams params) {
        // Add all known protocols to the repository.
        maxPendingCount = params.getMaxPendingCount();
        maxPendingSize  = params.getMaxPendingSize();
        for (int i = 0, len = params.getNumProtocols(); i < len; ++i) {
            protocolRepository.putProtocol(params.getProtocol(i));

            if (params.getProtocol(i).getMetrics() != null) {
                metrics.protocols.addMetric(params.getProtocol(i).getMetrics());
            }
        }

        // Attach and start network.
        this.net = net;
        net.attach(this);
        if ( ! net.waitUntilReady(120))
            throw new IllegalStateException("Network failed to become ready in time.");

        // Start messenger.
        msn = new Messenger();

        RetryPolicy retryPolicy = params.getRetryPolicy();
        if (retryPolicy != null) {
            resender = new Resender(retryPolicy);
            msn.addRecurrentTask(new ResenderTask(resender));
        } else {
            resender = null;
        }
        careTaker.setDaemon(true);
        careTaker.start();

        msn.start();
    }

    /**
     * <p>Returns the metrics used by this messagebus.</p>
     *
     * @return The metric set.
     */
    public MessageBusMetricSet getMetrics() {
        return metrics;
    }

    /**
     * <p>Sets the destroyed flag to true. The very first time this method is
     * called, it cleans up all its dependencies. Even if you retain a reference
     * to this object, all of its content is allowed to be garbage
     * collected.</p>
     *
     * @return True if content existed and was destroyed.
     */
    public boolean destroy() {
        if (!destroyed.getAndSet(true)) {
            try {
                careTaker.join();
            } catch (InterruptedException e) { }
            protocolRepository.clearPolicyCache();
            net.shutdown();
            msn.destroy();
            if (resender != null) {
                resender.destroy();
            }
            return true;
        }
        return false;
    }

    /**
     * <p>Synchronize with internal threads. This method will handshake with all
     * internal threads. This has the implicit effect of waiting for all active
     * callbacks. Note that this method should never be invoked from a callback
     * since that would make the thread wait for itself... forever. This method
     * is typically used to untangle during session shutdown.</p>
     */
    public void sync() {
        msn.sync();
        net.sync();
    }

    /**
     * <p>This is a convenience method to call {@link
     * #createSourceSession(SourceSessionParams)} with default values for the
     * {@link SourceSessionParams} object.</p>
     *
     * @param handler The reply handler to receive the replies for the session.
     * @return The created session.
     */
    public SourceSession createSourceSession(ReplyHandler handler) {
        return createSourceSession(new SourceSessionParams().setReplyHandler(handler));
    }

    /**
     * <p>This is a convenience method to call {@link
     * #createSourceSession(SourceSessionParams)} by first assigning the reply
     * handler to the parameter object.</p>
     *
     * @param handler The reply handler to receive the replies for the session.
     * @param params  The parameters to control the session.
     * @return The created session.
     */
    public SourceSession createSourceSession(ReplyHandler handler, SourceSessionParams params) {
        return createSourceSession(new SourceSessionParams(params).setReplyHandler(handler));
    }

    /**
     * <p>Creates a source session on top of this message bus.</p>
     *
     * @param params The parameters to control the session.
     * @return The created session.
     */
    public SourceSession createSourceSession(SourceSessionParams params) {
        if (destroyed.get()) {
            throw new IllegalStateException("Object is destroyed.");
        }
        return new SourceSession(this, params);
    }

    /**
     * <p>This is a convenience method to call {@link
     * #createIntermediateSession(IntermediateSessionParams)} with default
     * values for the {@link IntermediateSessionParams} object.</p>
     *
     * @param name          The local unique name for the created session.
     * @param broadcastName Whether or not to broadcast this session's name on
     *                      the network.
     * @param msgHandler    The handler to receive the messages for the session.
     * @param replyHandler  The handler to received the replies for the session.
     * @return The created session.
     */
    public IntermediateSession createIntermediateSession(String name,
                                                         boolean broadcastName,
                                                         MessageHandler msgHandler,
                                                         ReplyHandler replyHandler) {
        return createIntermediateSession(
                new IntermediateSessionParams()
                        .setName(name)
                        .setBroadcastName(broadcastName)
                        .setMessageHandler(msgHandler)
                        .setReplyHandler(replyHandler));
    }

    /**
     * <p>Creates an intermediate session on top of this message bus using the
     * given handlers and parameter object.</p>
     *
     * @param params The parameters to control the session.
     * @return The created session.
     */
    public synchronized IntermediateSession createIntermediateSession(IntermediateSessionParams params) {
        if (destroyed.get()) {
            throw new IllegalStateException("Object is destroyed.");
        }
        if (sessions.containsKey(params.getName())) {
            throw new IllegalArgumentException("Name '" + params.getName() + "' is not unique.");
        }
        IntermediateSession session = new IntermediateSession(this, params);
        sessions.put(params.getName(), session);
        if (params.getBroadcastName()) {
            net.registerSession(params.getName());
        }
        return session;
    }

    /**
     * <p>This is a convenience method to call {@link
     * #createDestinationSession(DestinationSessionParams)} with default values
     * for the {@link DestinationSessionParams} object.</p>
     *
     * @param name          The local unique name for the created session.
     * @param broadcastName Whether or not to broadcast this session's name on
     *                      the network.
     * @param handler       The handler to receive the messages for the session.
     * @return The created session.
     */
    public DestinationSession createDestinationSession(String name,
                                                       boolean broadcastName,
                                                       MessageHandler handler) {
        return createDestinationSession(
                new DestinationSessionParams()
                        .setName(name)
                        .setBroadcastName(broadcastName)
                        .setMessageHandler(handler));
    }

    /**
     * <p>Creates a destination session on top of this message bus using the
     * given handlers and parameter object.</p>
     *
     * @param params The parameters to control the session.
     * @return The created session.
     */
    public synchronized DestinationSession createDestinationSession(DestinationSessionParams params) {
        if (destroyed.get()) {
            throw new IllegalStateException("Object is destroyed.");
        }
        if (sessions.containsKey(params.getName())) {
            throw new IllegalArgumentException("Name '" + params.getName() + "' is not unique.");
        }
        DestinationSession session = new DestinationSession(this, params);
        sessions.put(params.getName(), session);
        if (params.getBroadcastName()) {
            net.registerSession(params.getName());
        }
        return session;
    }

    /**
     * <p>This method is invoked by the {@link
     * com.yahoo.messagebus.IntermediateSession#destroy()} to unregister
     * sessions from receiving data from message bus.</p>
     *
     * @param name          The name of the session to remove.
     * @param broadcastName Whether or not session name was broadcast.
     */
    public synchronized void unregisterSession(String name, boolean broadcastName) {
        if (broadcastName) {
            net.unregisterSession(name);
        }
        sessions.remove(name);
    }

    private boolean doAccounting() {
        return (maxPendingCount > 0 || maxPendingSize > 0);
    }
    /**
     * <p>This method handles choking input data so that message bus does not
     * blindly accept everything. This prevents an application running
     * out-of-memory in case it fail to choke input data itself. If this method
     * returns false, it means that it should be rejected.</p>
     *
     * @param msg The message to count.
     * @return True if the message was accepted.
     */
    private boolean checkPending(Message msg) {
        boolean busy = false;
        int size = msg.getApproxSize();

        if (doAccounting()) {
            synchronized (this) {
                busy = ((maxPendingCount > 0 && pendingCount >= maxPendingCount) ||
                        (maxPendingSize > 0 && pendingSize >= maxPendingSize));
                if (!busy) {
                    pendingCount++;
                    pendingSize += size;
                }
            }
        }
        if (busy) {
            return false;
        }
        msg.setContext(size);
        msg.pushHandler(this);
        return true;
    }

    @Override
    public void handleMessage(Message msg) {
        if (resender != null && msg.hasBucketSequence()) {
            deliverError(msg, ErrorCode.SEQUENCE_ERROR, "Bucket sequences not supported when resender is enabled.");
            return;
        }
        SendProxy proxy = new SendProxy(this, net, resender);
        msn.deliverMessage(msg, proxy);
    }

    @Override
    public void handleReply(Reply reply) {
        if (destroyed.get()) {
            reply.discard();
            return;
        }
        if (doAccounting()) {
            synchronized (this) {
                --pendingCount;
                pendingSize -= (Integer)reply.getContext();
            }
        }
        deliverReply(reply, reply.popHandler());
    }

    @Override
    public void deliverMessage(Message msg, String session) {
        MessageHandler msgHandler = sessions.get(session);
        if (msgHandler == null) {
            deliverError(msg, ErrorCode.UNKNOWN_SESSION, "Session '" + session + "' does not exist.");
        } else if (!checkPending(msg)) {
            deliverError(msg, ErrorCode.SESSION_BUSY, "Session '" + net.getConnectionSpec() + "/" + session +
                                                      "' is busy, try again later.");
        } else {
            msn.deliverMessage(msg, msgHandler);
        }
    }

    /**
     * <p>Adds a protocol to the internal repository of protocols, replacing any
     * previous instance of the protocol and clearing the associated routing
     * policy cache.</p>
     *
     * @param protocol The protocol to add.
     */
    public void putProtocol(Protocol protocol) {
        protocolRepository.putProtocol(protocol);
    }

    @Override
    public Protocol getProtocol(Utf8Array name) {
        return protocolRepository.getProtocol(name.toString());
    }

    public Protocol getProtocol(Utf8String name) {
        return getProtocol((Utf8Array)name);
    }

    @Override
    public void deliverReply(Reply reply, ReplyHandler handler) {
        msn.deliverReply(reply, handler);
    }

    @Override
    public void setupRouting(RoutingSpec spec) {
        Map<String, RoutingTable> tables = new HashMap<String, RoutingTable>();
        for (int i = 0, len = spec.getNumTables(); i < len; ++i) {
            RoutingTableSpec table = spec.getTable(i);
            String name = table.getProtocol();
            if (!protocolRepository.hasProtocol(name)) {
                log.log(LogLevel.INFO, "Protocol '" + name + "' is not supported, ignoring routing table.");
                continue;
            }
            tables.put(name, new RoutingTable(table));
        }
        tablesRef.set(tables);
        protocolRepository.clearPolicyCache();
    }

    /**
     * <p>Returns the resender that is running within this message bus.</p>
     *
     * @return The resender.
     */
    public Resender getResender() {
        return resender;
    }

    /**
     * <p>Returns the number of messages received that have not been replied to
     * yet.</p>
     *
     * @return The pending count.
     */
    public synchronized int getPendingCount() {
        return pendingCount;
    }

    /**
     * <p>Returns the size of messages received that have not been replied to
     * yet.</p>
     *
     * @return The pending size.
     */
    public synchronized int getPendingSize() {
        return pendingSize;
    }

    /**
     * <p>Sets the maximum number of messages that can be received without being
     * replied to yet.</p>
     *
     * @param maxCount The max count.
     */
    public void setMaxPendingCount(int maxCount) {
        maxPendingCount = maxCount;
    }

    /**
     * Gets maximum number of messages that can be received without being
     * replied to yet.
     */
    public int getMaxPendingCount() {
        return maxPendingCount;
    }

    /**
     * <p>Sets the maximum size of messages that can be received without being
     * replied to yet.</p>
     *
     * @param maxSize The max size.
     */
    public void setMaxPendingSize(int maxSize) {
        maxPendingSize = maxSize;
    }

    /**
     * Gets maximum combined size of messages that can be received without
     * being replied to yet.
     */
    public int getMaxPendingSize() {
        return maxPendingSize;
    }

    /**
     * <p>Returns a named routing table, may return null.</p>
     *
     * @param name The name of the routing table to return.
     * @return The routing table object.
     */
    public RoutingTable getRoutingTable(String name) {
        Map<String, RoutingTable> tables = tablesRef.get();
        if (tables == null) {
            return null;
        }
        return tables.get(name);
    }
    /**
     * <p>Returns a named routing table, may return null.</p>
     *
     * @param name The name of the routing table to return.
     * @return The routing table object.
     */
    public RoutingTable getRoutingTable(Utf8String name) {

        return getRoutingTable(name.toString());
    }

    /**
     * <p>Returns a routing policy that corresponds to the argument protocol
     * name, policy name and policy parameter. This will cache reuse all
     * policies as soon as they are first requested.</p>
     *
     * @param protocolName The name of the protocol to invoke {@link Protocol#createPolicy(String,String)} on.
     * @param policyName   The name of the routing policy to retrieve.
     * @param policyParam  The parameter for the routing policy to retrieve.
     * @return A corresponding routing policy, or null.
     */
    public RoutingPolicy getRoutingPolicy(String protocolName, String policyName, String policyParam) {
        return protocolRepository.getRoutingPolicy(protocolName, policyName, policyParam);
    }

    /**
     * <p>Returns a routing policy that corresponds to the argument protocol
     * name, policy name and policy parameter. This will cache reuse all
     * policies as soon as they are first requested.</p>
     *
     * @param protocolName The name of the protocol to invoke {@link Protocol#createPolicy(String,String)} on.
     * @param policyName   The name of the routing policy to retrieve.
     * @param policyParam  The parameter for the routing policy to retrieve.
     * @return A corresponding routing policy, or null.
     */
    public RoutingPolicy getRoutingPolicy(Utf8String protocolName, String policyName, String policyParam) {
        return protocolRepository.getRoutingPolicy(protocolName.toString(), policyName, policyParam);
    }

    /**
     * <p>Returns the connection spec string for the network layer of this
     * message bus. This is merely a proxy of the same function in the network
     * layer.</p>
     *
     * @return The connection string.
     */
    public String getConnectionSpec() {
        return net.getConnectionSpec();
    }

    /**
     * <p>Constructs and schedules a Reply containing an error to the handler of the given Message.</p>
     *
     * @param msg     The message to reply to.
     * @param errCode The code of the error to set.
     * @param errMsg  The message of the error to set.
     */
    private void deliverError(Message msg, int errCode, String errMsg) {
        Reply reply = new EmptyReply();
        reply.swapState(msg);
        reply.addError(new Error(errCode, errMsg));
        deliverReply(reply, reply.popHandler());
    }

    /**
     * <p>Implements a task for running the resender in the messenger
     * thread. This task acts as a proxy for the resender, allowing the task to
     * be deleted without affecting the resender itself.</p>
     */
    private static class ResenderTask implements Messenger.Task {

        final Resender resender;

        ResenderTask(Resender resender) {
            this.resender = resender;
        }

        public void destroy() {
            // empty
        }

        public void run() {
            resender.resendScheduled();
        }

    }
}

File: messagebus/src/vespa/messagebus/messagebus.cpp
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "messagebus.h"
#include "messenger.h"
#include "emptyreply.h"
#include "errorcode.h"
#include "sendproxy.h"
#include "protocolrepository.h"
#include <vespa/messagebus/network/inetwork.h>
#include <vespa/vespalib/util/exceptions.h>
#include <vespa/vespalib/util/gate.h>

#include <vespa/log/log.h>
LOG_SETUP(".messagebus");

using vespalib::make_string;
using namespace std::chrono_literals;

namespace {

/**
 * Implements a task for running the resender in the messenger thread. This task
 * acts as a proxy for the resender, allowing the task to be deleted without
 * affecting the resender itself.
 */
class ResenderTask : public mbus::Messenger::ITask {
private:
    mbus::Resender *_resender;

public:
    ResenderTask(mbus::Resender &resender)
        : _resender(&resender)
    {
        // empty
    }

    void run() override {
        _resender->resendScheduled();
    }

    uint8_t priority() const override {
        return 255;
    }
};

/**
 * Implements a task for monitoring shutdown of the messenger thread. This task
 * helps to determine whether or not there is any work left in either the
 * messenger or network thread.
 */
class ShutdownTask : public mbus::Messenger::ITask {
private:
    mbus::INetwork  &_net;
    mbus::Messenger &_msn;
    bool            &_done;
    vespalib::Gate  &_gate;

public:
    ShutdownTask(mbus::INetwork &net, mbus::Messenger &msn,
                 bool &done, vespalib::Gate &gate)
        : _net(net),
          _msn(msn),
          _done(done),
          _gate(gate)
    { }

    ~ShutdownTask() override {
        _gate.countDown();
    }

    void run() override {
        _net.postShutdownHook();
        _done = _msn.isEmpty();
    }

    uint8_t priority() const override {
        return 255;
    }
};

} // anonymous

namespace mbus {

MessageBus::MessageBus(INetwork &net, ProtocolSet protocols) :
    _network(net),
    _lock(),
    _routingTables(),
    _sessions(),
    _protocolRepository(std::make_unique<ProtocolRepository>()),
    _msn(std::make_unique<Messenger>()),
    _resender(),
    _maxPendingCount(0),
    _maxPendingSize(0),
    _pendingCount(0),
    _pendingSize(0)
{
    MessageBusParams params;
    while (!protocols.empty()) {
        IProtocol::SP protocol = protocols.extract();
        if (protocol) {
            params.addProtocol(protocol);
        }
    }
    setup(params);
}

MessageBus::MessageBus(INetwork &net, const MessageBusParams &params) :
    _network(net),
    _lock(),
    _routingTables(),
    _sessions(),
    _protocolRepository(std::make_unique<ProtocolRepository>()),
    _msn(std::make_unique<Messenger>()),
    _resender(),
    _maxPendingCount(params.getMaxPendingCount()),
    _maxPendingSize(params.getMaxPendingSize()),
    _pendingCount(0),
    _pendingSize(0)
{
    setup(params);
}

MessageBus::~MessageBus()
{
    // all sessions must have been destroyed prior to this,
    // so no more traffic from clients
    _msn->discardRecurrentTasks(); // no more traffic from recurrent tasks
    _network.shutdown(); // no more traffic from network

    bool done = false;
    while (!done) {
        vespalib::Gate gate;
        _msn->enqueue(std::make_unique<ShutdownTask>(_network, *_msn, done, gate));
        gate.await();
    }
}

void
MessageBus::setup(const MessageBusParams &params)
{
    // Add all known protocols to the repository.
    for (uint32_t i = 0, len = params.getNumProtocols(); i < len; ++i) {
        _protocolRepository->putProtocol(params.getProtocol(i));
    }

    // Attach and start network.
    _network.attach(*this);
    if (!_network.start()) {
        throw vespalib::NetworkSetupFailureException("Failed to start network.");
    }
    if (!_network.waitUntilReady(120s)) {
        throw vespalib::NetworkSetupFailureException("Network failed to become ready in time.");
    }

    // Start messenger.
    IRetryPolicy::SP retryPolicy = params.getRetryPolicy();
    if (retryPolicy) {
        _resender = std::make_unique<Resender>(retryPolicy);

        _msn->addRecurrentTask(std::make_unique<ResenderTask>(*_resender));
    }
    if (!_msn->start()) {
        throw vespalib::NetworkSetupFailureException("Failed to start messenger.");
    }
}

SourceSession::UP
MessageBus::createSourceSession(IReplyHandler &handler)
{
    return createSourceSession(SourceSessionParams().setReplyHandler(handler));
}

SourceSession::UP
MessageBus::createSourceSession(IReplyHandler &handler,
                                const SourceSessionParams &params)
{
    return createSourceSession(SourceSessionParams(params).setReplyHandler(handler));
}

SourceSession::UP
MessageBus::createSourceSession(const SourceSessionParams &params)
{
    return SourceSession::UP(new SourceSession(*this, params));
}

IntermediateSession::UP
MessageBus::createIntermediateSession(const string &name,
                                      bool broadcastName,
                                      IMessageHandler &msgHandler,
                                      IReplyHandler &replyHandler)
{
    return createIntermediateSession(IntermediateSessionParams()
                                     .setName(name)
                                     .setBroadcastName(broadcastName)
                                     .setMessageHandler(msgHandler)
                                     .setReplyHandler(replyHandler));
}

IntermediateSession::UP
MessageBus::createIntermediateSession(const IntermediateSessionParams &params)
{
    std::lock_guard guard(_lock);
    IntermediateSession::UP ret(new IntermediateSession(*this, params));
    _sessions[params.getName()] = ret.get();
    if (params.getBroadcastName()) {
        _network.registerSession(params.getName());
    }
    return ret;
}

DestinationSession::UP
MessageBus::createDestinationSession(const string &name,
                                     bool broadcastName,
                                     IMessageHandler &handler)
{
    return createDestinationSession(DestinationSessionParams()
                                    .setName(name)
                                    .setBroadcastName(broadcastName)
                                    .setMessageHandler(handler));
}

DestinationSession::UP
MessageBus::createDestinationSession(const DestinationSessionParams &params)
{
    std::lock_guard guard(_lock);
    DestinationSession::UP ret(new DestinationSession(*this, params));
    _sessions[params.getName()] = ret.get();
    if (params.getBroadcastName()) {
        _network.registerSession(params.getName());
    }
    return ret;
}

void
MessageBus::unregisterSession(const string &sessionName)
{
    std::lock_guard guard(_lock);
    _network.unregisterSession(sessionName);
    _sessions.erase(sessionName);
}

RoutingTable::SP
MessageBus::getRoutingTable(const string &protocol)
{
    typedef std::map<string, RoutingTable::SP>::iterator ITR;
    std::lock_guard guard(_lock);
    ITR itr = _routingTables.find(protocol);
    if (itr == _routingTables.end()) {
        return RoutingTable::SP(); // not found
    }
    return itr->second;
}

IRoutingPolicy::SP
MessageBus::getRoutingPolicy(const string &protocolName,
                             const string &policyName,
                             const string &policyParam)
{
    return _protocolRepository->getRoutingPolicy(protocolName, policyName, policyParam);
}

void
MessageBus::sync()
{
    _msn->sync();
    _network.sync(); // should not be necessary, as msn is intermediate
}

void
MessageBus::handleMessage(Message::UP msg)
{
    if (_resender && msg->hasBucketSequence()) {
        deliverError(std::move(msg), ErrorCode::SEQUENCE_ERROR,
                     "Bucket sequences not supported when resender is enabled.");
        return;
    }
    SendProxy &proxy = *(new SendProxy(*this, _network, _resender.get())); // deletes self
    _msn->deliverMessage(std::move(msg), proxy);
}

bool
MessageBus::setupRouting(const RoutingSpec &spec)
{
    std::map<string, RoutingTable::SP> rtm;
    for (uint32_t i = 0; i < spec.getNumTables(); ++i) {
        const RoutingTableSpec &cfg = spec.getTable(i);
        if (getProtocol(cfg.getProtocol()) == nullptr) { // protocol not found
            LOG(info, "Protocol '%s' is not supported, ignoring routing table.", cfg.getProtocol().c_str());
            continue;
        }
        rtm[cfg.getProtocol()] = std::make_shared<RoutingTable>(cfg);
    }
    {
        std::lock_guard guard(_lock);
        std::swap(_routingTables, rtm);
    }
    _protocolRepository->clearPolicyCache();
    return true;
}

IProtocol *
MessageBus::getProtocol(const string &name)
{
    return _protocolRepository->getProtocol(name);
}

IProtocol::SP
MessageBus::putProtocol(const IProtocol::SP & protocol)
{
    return _protocolRepository->putProtocol(protocol);
}

bool
MessageBus::checkPending(Message &msg)
{
    bool busy = false;
    const uint32_t size = msg.getApproxSize();
    {
        constexpr auto relaxed = std::memory_order_relaxed;
        const uint32_t maxCount = _maxPendingCount.load(relaxed);
        const uint32_t maxSize = _maxPendingSize.load(relaxed);
        if (maxCount > 0 || maxSize > 0) {
            busy = ((maxCount > 0 && _pendingCount.load(relaxed) >= maxCount) ||
                    (maxSize > 0 && _pendingSize.load(relaxed) >= maxSize));
            if (!busy) {
                _pendingCount.fetch_add(1, relaxed);
                _pendingSize.fetch_add(size, relaxed);
            }
        }
    }
    if (busy) {
        return false;
    }
    msg.setContext(Context(static_cast<uint64_t>(size)));
    msg.pushHandler(*this, *this);
    return true;
}

void
MessageBus::handleReply(Reply::UP reply)
{
    _pendingCount.fetch_sub(1, std::memory_order_relaxed);
    _pendingSize.fetch_sub(reply->getContext().value.UINT64,
                           std::memory_order_relaxed);
    IReplyHandler &handler = reply->getCallStack().pop(*reply);
    deliverReply(std::move(reply), handler);
}

void
MessageBus::handleDiscard(Context ctx)
{
    _pendingCount.fetch_sub(1, std::memory_order_relaxed);
    _pendingSize.fetch_sub(ctx.value.UINT64, std::memory_order_relaxed);
}

void
MessageBus::deliverMessage(Message::UP msg, const string &session)
{
    IMessageHandler *msgHandler = nullptr;
    {
        std::lock_guard guard(_lock);
        std::map<string, IMessageHandler*>::iterator it = _sessions.find(session);
        if (it != _sessions.end()) {
            msgHandler = it->second;
        }
    }
    if (msgHandler == nullptr) {
        deliverError(std::move(msg), ErrorCode::UNKNOWN_SESSION,
                     make_string("Session '%s' does not exist.", session.c_str()));
    } else if (!checkPending(*msg)) {
        deliverError(std::move(msg), ErrorCode::SESSION_BUSY,
                     make_string("Session '%s' is busy, try again later.", session.c_str()));
    } else {
        _msn->deliverMessage(std::move(msg), *msgHandler);
    }
}

void
MessageBus::deliverError(Message::UP msg, uint32_t errCode, const string &errMsg)
{
    auto reply = std::make_unique<EmptyReply>();
    reply->swapState(*msg);
    reply->addError(Error(errCode, errMsg));

    IReplyHandler &replyHandler = reply->getCallStack().pop(*reply);
    deliverReply(std::move(reply), replyHandler);
}

void
MessageBus::deliverReply(Reply::UP reply, IReplyHandler &handler)
{
    _msn->deliverReply(std::move(reply), handler);
}

string
MessageBus::getConnectionSpec() const
{
    return _network.getConnectionSpec();
}

void
MessageBus::setMaxPendingCount(uint32_t maxCount)
{
    _maxPendingCount.store(maxCount, std::memory_order_relaxed);
}

void
MessageBus::setMaxPendingSize(uint32_t maxSize)
{
    _maxPendingSize.store(maxSize, std::memory_order_relaxed);
}

} // namespace mbus
